platform/android/MapboxGLAndroidSDK/src/main/java/com/mapbox/mapboxsdk/maps/AnnotationManager.java

```diff
@@ -329,84 +329,131 @@ void adjustTopOffsetPixels(MapboxMap mapboxMap) {
   // Click event
   //
 
-  boolean onTap(PointF tapPoint, float screenDensity) {
-    float toleranceSides = 4 * screenDensity;
-    float toleranceTopBottom = 10 * screenDensity;
-    boolean handledDefaultClick = false;
-
-    RectF tapRect = new RectF(tapPoint.x - iconManager.getAverageIconWidth() / 2 - toleranceSides,
-      tapPoint.y - iconManager.getAverageIconHeight() / 2 - toleranceTopBottom,
-      tapPoint.x + iconManager.getAverageIconWidth() / 2 + toleranceSides,
-      tapPoint.y + iconManager.getAverageIconHeight() / 2 + toleranceTopBottom);
-
-    List<Marker> nearbyMarkers = getMarkersInRect(tapRect);
-    long newSelectedMarkerId = -1;
-
-    // find a Marker that isn't selected yet
-    if (nearbyMarkers.size() > 0) {
-      Collections.sort(nearbyMarkers);
-      for (Marker nearbyMarker : nearbyMarkers) {
-        boolean found = false;
-        for (Marker selectedMarker : selectedMarkers) {
-          if (selectedMarker.equals(nearbyMarker)) {
-            found = true;
-          }
-        }
-        if (!found) {
-          newSelectedMarkerId = nearbyMarker.getId();
-          break;
+  boolean onTap(PointF tapPoint) {
+    MarkerHit markerHit = getMarkerHitFromTouchArea(tapPoint);
+    long markerId = new MarkerHitResolver(markerViewManager, mapboxMap.getProjection()).execute(markerHit);
+    return markerId >= 0 && isClickHandledForMarker(markerId);
+  }
+
+  private MarkerHit getMarkerHitFromTouchArea(PointF tapPoint) {
+    int averageIconWidthOffset = iconManager.getAverageIconWidth() / 2;
+    int averageIconHeightOffset = iconManager.getAverageIconHeight() / 2;
+    final RectF tapRect = new RectF(tapPoint.x - averageIconWidthOffset,
+      tapPoint.y - averageIconHeightOffset,
+      tapPoint.x + averageIconWidthOffset,
+      tapPoint.y + averageIconHeightOffset
+    );
+    return new MarkerHit(tapRect, getMarkersInRect(tapRect));
+  }
+
+  private boolean isClickHandledForMarker(long markerId) {
+    boolean handledDefaultClick;
+    Marker marker = (Marker) getAnnotation(markerId);
+    if (marker instanceof MarkerView) {
```

Comment:
good idea, would love to get in https://github.com/mapbox/mapbox-gl-native/issues/6912 first as this will changes the whole setup around annotations (this makes 50% of our current java code around annotations obsolete) and would like to avoid doing twice the work.